docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,479 @@
+operator from the sortition pool or reducing its weight, changing the operator's
+eligibility for rewards, and so forth.
+
+Every application callback executed as a result of slash should have a 250k gas
+limit. Slashing are processed in a FIFO basis, and there is just one function
+exposed by the staking contract allowing to slash one or more operators from the
+head of the queue. Callback failure does not revert the transaction.
+
+In the case of legacy stakers, their liquid T is slashed first before a call to 
+the legacy contract is executed.
+
+It is important to note slashing executed in the context of one application may
+lead to involuntarily decreasing the authorization for other applications in 
+case the amount of stake available after the slashing is lower than these
+authorizations.
+
+=== Limitations
+
+An absolute minimum is that existing KEEP/NU stakes should be supported in T
+network for Random Beacon, TBTC v2, and Proxy Re-Encryption. Given the fact
+supporting legacy KEEP/NU stakes forever, in every other application implemented
+for T, might not be always possible, it is expected that applications are able
+to revert authorization transactions if the delegation used in the authorization
+order does not meet the application requirements.
+
+=== Upgradeability
+
+The staking contract will be upgradeable. The exact upgradeability mechanism is
+out of the scope of this document.
+
+== Public API
+
+=== Delegating a stake
+
+==== `stake(address operator, address beneficiary, address authorizer, uint256 amount) external` 
+   
+Creates a delegation with `msg.sender` owner with the given operator,
+beneficiary, and authorizer. Transfers the given amount of T to the staking
+contract. The owner of the delegation needs to have the amount approved to
+transfer to the staking contract.
+
+==== `stakeKeep(address operator) external`
+
+Copies delegation from the legacy KEEP staking contract to T staking contract.
+No tokens are transferred. Caches the active stake amount from KEEP staking 
+contract. Can be called by anyone.
+
+==== `stakeNu(address operator, address beneficiary, address authorizer) external`
+
+Copies delegation from the legacy NU staking contract to T staking contract,
+additionally appointing beneficiary and authorizer roles. Caches the amount
+staked in NU staking contract. Can be called only by the original delegation
+owner.
+
+==== `setMinimumStakeAmount(uint256 amount) external onlyGovernance`
+
+Allows the governance to set the minimum required stake amount. This amount is
+required to protect against griefing the staking contract and individual
+applications are allowed to require higher minimum stakes if necessary.  
+
+=== Authorizing an application
+
+==== `approveApplication(address application) external onlyGovernance`
+
+Allows the governance to approve the particular application before individual
+stake authorizers are able to authorize it.
```

Comment:
Do we want to have mapping or list in input parameters to specify which stakes are approved (T, KEEP, NU or all of them)?

Reply:
I am definitely for simplicity and not worrying about availability on start. 
Just think that all logic could be included in staking contract and not so big. Because when app is not available for some legacy stake that means when authorization increased - ceil of authorization is lower (same for slashing/syncing). App would work only with authorization knowing nothing about the type of stake. 
But again I totally agree not to implement this on day one

Reply:
To explain the problematic case I have in mind:
- Operator has 100k KEEP and 1M T staked. For simplicity, let's assume 1 KEEP = 1 T. In total, it's 1.1M `amount` and applications has no idea what is the composition of tokens (applications shouldn't care about it). Now, should we let this operator to authorize application X to `amount = 100k` if it refuses to work with KEEP staking contract? Yes, because the operator has 1M T. So the authorization for that particular application could be for 1M max. 
- Now, that operator calls `unstakeT(amount = 1M)`. We will notify the application `amount` changed, and that it's 100k now but that application is not aware that amount is only in KEEP.

What should happen in step 2 is that the application should revert the TX. But it means applications need to be aware of the composition of stakes. Alternatively, the staking contract would need to check applications, see what stakes they are authorized for, and update only in specific circumstances.

This is getting quite complex in my opinion and T staking contract is already pretty complex. I am afraid, we could be designing for a situation that will never happen given that applications for the next few years could be perfectly fine with 5% notifier reward limit. If we agree on respecting this limit, we are eliminating some complexity from the implementation and sending a friendly signal to existing Keep stakers who cannot unstake from tBTC v1 right now.

Reply:
I see that in `unstakeT` won't be calls to app in anyway because authorization remains the same.
Hypothetical how it could be implemented in staking contract v2: When we will calculate how many tokens reallly can be unstaked - we can see that amount for this app was 100k from 1M possible which means only 900k can be unstaked safely. So app knows nothing, staking contract reverts tx

Reply:
> I see that in unstakeT won't be calls to app in anyway because authorization remains the same.

Yes, not in the current version of the RFC. I meant that we could add it - just one of the options.

> Hypotheticaly how it could be implemented in staking contract v2: When we will calculate how many tokens reallly can be unstaked - we can see that amount for this app was 100k from 1M possible which means only 900k can be unstaked safely. So app knows nothing, staking contract reverts tx
 
Right, but in this case, the staking contract needs to know what type of stake the given application is willing to accept and validate the current composition each time `unstake*` happens. It is definitely possible to implement, I am just concerned if we really need it given the contract is upgradeable.

Reply:
> Yes, not in the current version of the RFC. I meant that we could add it - just one of the options.

Could you elaborate? In current version decreasing authorization and unstaking are not possible in one tx (because of backward call from app). Or you mean something else?

> Right, but in this case, the staking contract needs to know what type of stake the given application is willing to accept and validate the current composition each time `unstake*` happens. It is definitely possible to implement, I am just concerned if we really need it given the contract is upgradeable.

👍

Reply:
> Could you elaborate? In current version decreasing authorization and unstaking are not possible in one tx (because of backward call from app). Or you mean something else?

I meant that if we want to have the application aware of the stake type, we could add one more callback to `unstake` to give the application a chance to take a look at the current stake composition and revert the transactions. I hate this idea but this was one of the options I had in mind so posted it here and... caused confusion. 

It seems we agree on starting simple, so let's just forget about this crazy idea of an application callback from `unstake`.

Reply:
Anyway, feels that we are aligned on everything so I'll work on updating the RFC to a final-ish form tomorrow!

Reply:
Another question: looks like different apps can have different minimum authorization (>= min stake). maybe add to `approve` logic also caching this parameter? Or maybe other approach?

Reply:
Not sure if I follow. According to the RFC, this is currently handled in the callback that is called for increase and decrease operations and that callback can revert the transaction if some conditions are not met. How would the change you described look like?

Reply:
Nothing special, just cache min amount per app in staking contract and check this beforehand during authorization. Not necessary of course